File: lib/core_dsp.dart
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:ffmpeg_kit_flutter_audio/ffmpeg_kit.dart';
import 'package:ffmpeg_kit_flutter_audio/return_code.dart';

class DspEngine {
  /// M2: AutoTrim - Recorte de silencios y ruido analógico
  Future<void> autoTrim(String filePath) async {
    final file = File(filePath);
    if (!file.existsSync()) throw Exception("Archivo no encontrado en I/O.");

    final tempFilePath =
        "${file.parent.path}${Platform.pathSeparator}temp_dsp_trim.mp3";
    final afFilter =
        "silenceremove=start_periods=1:start_threshold=-35dB,areverse,silenceremove=start_periods=1:start_threshold=-35dB,areverse";

    debugPrint("⏳ [M2] Aplicando Auto-Trim: ${file.uri.pathSegments.last}");
    bool success = await _runFFmpeg(filePath, tempFilePath, afFilter);

    if (success) {
      await _atomicReplace(file, File(tempFilePath));
      debugPrint("🟢 [M2 ÉXITO] Archivo truncado.");
    }
  }

  /// M3: Normalización LUFS de Doble Pasada (Dual-Pass EBU R128)
  Future<void> normalizeLUFS(String filePath) async {
    final file = File(filePath);
    if (!file.existsSync()) throw Exception("Archivo no encontrado en I/O.");

    debugPrint(
      "⏳ [M3] Iniciando Doble Pasada LUFS: ${file.uri.pathSegments.last}",
    );

    String? i, lra, tp, thresh;

    // PASO 1: Análisis Diferencial
    if (Platform.isWindows || Platform.isLinux) {
      final pass1Result = await Process.run('ffmpeg', [
        '-i',
        filePath,
        '-af',
        'loudnorm=I=-14:LRA=11:TP=-1.5:print_format=json',
        '-f',
        'null',
        'NUL',
      ]);
      final log = pass1Result.stderr.toString();
      i = RegExp(r'"input_i" : "([-0-9.]+)"').firstMatch(log)?.group(1);
      lra = RegExp(r'"input_lra" : "([-0-9.]+)"').firstMatch(log)?.group(1);
      tp = RegExp(r'"input_tp" : "([-0-9.]+)"').firstMatch(log)?.group(1);
      thresh = RegExp(
        r'"input_thresh" : "([-0-9.]+)"',
      ).firstMatch(log)?.group(1);
    } else {
      final session = await FFmpegKit.execute(
        '-i "$filePath" -af loudnorm=I=-14:LRA=11:TP=-1.5:print_format=json -f null /dev/null',
      );
      final log = await session.getLogsAsString();
      i = RegExp(r'"input_i" : "([-0-9.]+)"').firstMatch(log)?.group(1);
      lra = RegExp(r'"input_lra" : "([-0-9.]+)"').firstMatch(log)?.group(1);
      tp = RegExp(r'"input_tp" : "([-0-9.]+)"').firstMatch(log)?.group(1);
      thresh = RegExp(
        r'"input_thresh" : "([-0-9.]+)"',
      ).firstMatch(log)?.group(1);
    }

    if (i == null || lra == null || tp == null || thresh == null) {
      debugPrint(
        "🔴 [M3 FATAL] Fallo en Paso 1: Extracción de matriz acústica fallida.",
      );
      return;
    }

    // PASO 2: Inyección de Ganancia
    final tempFilePath =
        "${file.parent.path}${Platform.pathSeparator}temp_dsp_norm.mp3";
    final pass2Filter =
        "loudnorm=I=-14:LRA=11:TP=-1.5:measured_I=$i:measured_LRA=$lra:measured_TP=$tp:measured_thresh=$thresh:linear=true";

    bool success = await _runFFmpeg(filePath, tempFilePath, pass2Filter);

    if (success) {
      await _atomicReplace(file, File(tempFilePath));
      debugPrint("🟢 [M3 ÉXITO] Volumen estandarizado a -14 LUFS.");
    }
  }

  /// PRE-CHECK: Valida si la pista ya pasó por el pipeline
  Future<bool> checkWatermark(String filePath) async {
    if (Platform.isWindows || Platform.isLinux) {
      final process = await Process.run('ffmpeg', [
        '-i',
        filePath,
        '-f',
        'ffmetadata',
        '-',
      ]);
      final log = process.stderr.toString() + process.stdout.toString();
      return log.contains('DjStudio_M3=Verified');
    } else {
      final session = await FFmpegKit.execute('-i "$filePath" -f ffmetadata -');
      final log = await session.getLogsAsString();
      return log.contains('DjStudio_M3=Verified');
    }
  }

  /// INYECCIÓN: Sella el archivo atómicamente
  Future<void> injectWatermark(String filePath) async {
    final file = File(filePath);
    final tempFilePath =
        "${file.parent.path}${Platform.pathSeparator}temp_watermark.mp3";

    final success = await _runFFmpeg(
      filePath,
      tempFilePath,
      "", // Sin filtro de audio
    );

    if (success) {
      await _atomicReplace(file, File(tempFilePath));
      debugPrint("🟢 [WATERMARK] Firma inyectada en ID3v2.");
    }
  }

  // ==========================================
  // CAPA DE INFRAESTRUCTURA (Helpers)
  // ==========================================
  Future<bool> _runFFmpeg(String input, String output, String filter) async {
    List<String> args;
    String cmdMobile;

    if (filter.isEmpty) {
      args = [
        '-y',
        '-i',
        input,
        '-map',
        '0',
        '-c',
        'copy',
        '-metadata',
        'DjStudio_M3=Verified',
        output,
      ];
      cmdMobile =
          '-y -i "$input" -map 0 -c copy -metadata DjStudio_M3=Verified "$output"';
    } else {
      args = [
        '-y',
        '-i',
        input,
        '-af',
        filter,
        '-c:a',
        'libmp3lame',
        '-q:a',
        '2',
        output,
      ];
      cmdMobile =
          '-y -i "$input" -af "$filter" -c:a libmp3lame -q:a 2 "$output"';
    }

    if (Platform.isWindows || Platform.isLinux) {
      final process = await Process.run('ffmpeg', args);
      if (process.exitCode != 0) {
        debugPrint("🔴 [FFMPEG NATIVO] ${process.stderr}");
      }
      return process.exitCode == 0;
    } else {
      final session = await FFmpegKit.execute(cmdMobile);
      final returnCode = await session.getReturnCode();
      if (!ReturnCode.isSuccess(returnCode)) {
        debugPrint("🔴 [FFMPEG MOBILE] ${await session.getLogsAsString()}");
      }
      return ReturnCode.isSuccess(returnCode);
    }
  }

  Future<void> _atomicReplace(File original, File temp) async {
    final originalPath = original.path;
    await original.delete();
    await temp.rename(originalPath);
  }
}

File: rust/src/api/core_dsp.rs
use std::fs;
use std::process::Command;
use std::path::Path;
use std::env;

#[flutter_rust_bridge::frb(init)]
pub fn init_app() {
    flutter_rust_bridge::setup_default_user_utils();
}

// 🛠️ RESOLUTOR DINÁMICO: Busca FFmpeg en la carpeta del ejecutable
fn get_ffmpeg_path() -> String {
    if let Ok(mut exe_path) = env::current_exe() {
        exe_path.pop();
        exe_path.push(if cfg!(target_os = "windows") { "ffmpeg.exe" } else { "ffmpeg" });
        if exe_path.exists() {
            return exe_path.to_string_lossy().into_owned();
        }
    }
    "ffmpeg".to_string()
}

fn extract_json_value(log: &str, key: &str) -> Option<String> {
    let search_key = format!("\"{}\" : \"", key);
    if let Some(start) = log.find(search_key.as_str()) {
        let val_start = start + search_key.len();
        if let Some(end_offset) = log[val_start..].find('\"') {
            return Some(log[val_start..val_start + end_offset].to_string());
        }
    }
    None
}

fn atomic_replace(temp_path: &Path, original_path: &Path) -> Result<(), String> {
    fs::rename(temp_path, original_path).map_err(|e| format!("Fallo en I/O Atómico: {}", e))
}

pub async fn process_auto_trim(input_path: String) -> Result<bool, String> {
    let input = Path::new(&input_path);
    if !input.exists() { return Err("Archivo no encontrado en I/O.".to_string()); }

    let temp_path = input.with_file_name("temp_dsp_trim.mp3");
    let filter = "silenceremove=start_periods=1:start_duration=0.05:start_threshold=-30dB,areverse,silenceremove=start_periods=1:start_duration=0.05:start_threshold=-30dB,areverse";
    
    let output = Command::new(get_ffmpeg_path())
        .args(["-y", "-i", input.to_str().unwrap(), "-af", filter, "-c:a", "libmp3lame", "-q:a", "2", temp_path.to_str().unwrap()])
        .output()
        .map_err(|e| format!("OS Invocation Error: {}", e))?;

    if output.status.success() {
        atomic_replace(&temp_path, input)?;
        Ok(true)
    } else {
        if temp_path.exists() { let _ = fs::remove_file(temp_path); }
        Err(String::from_utf8_lossy(&output.stderr).to_string())
    }
}

pub async fn inject_watermark(input_path: String) -> Result<bool, String> {
    let input = Path::new(&input_path);
    let temp_path = input.with_file_name("temp_watermark.mp3");

    let output = Command::new(get_ffmpeg_path())
        .args(["-y", "-i", input.to_str().unwrap(), "-map", "0", "-c", "copy", "-metadata", "DjStudio_M3_V2=Verified", temp_path.to_str().unwrap()])
        .output()
        .map_err(|e| format!("OS Invocation Error: {}", e))?;

    if output.status.success() {
        atomic_replace(&temp_path, input)?;
        Ok(true)
    } else {
        if temp_path.exists() { let _ = fs::remove_file(temp_path); }
        Err(String::from_utf8_lossy(&output.stderr).to_string())
    }
}

pub async fn check_watermark(input_path: String) -> Result<bool, String> {
    let output = Command::new(get_ffmpeg_path())
        .args(["-i", &input_path, "-f", "ffmetadata", "-"])
        .output()
        .map_err(|e| format!("OS Invocation Error: {}", e))?;

    let log = format!("{}{}", String::from_utf8_lossy(&output.stderr), String::from_utf8_lossy(&output.stdout));
    Ok(log.contains("DjStudio_M3_V2=Verified"))
}

pub async fn normalize_lufs(input_path: String) -> Result<bool, String> {
    let input = Path::new(&input_path);
    if !input.exists() { return Err("Archivo no encontrado en I/O.".to_string()); }

    let null_sink = if cfg!(target_os = "windows") { "NUL" } else { "/dev/null" };

    let pass1_output = Command::new(get_ffmpeg_path())
        .args(["-i", input.to_str().unwrap(), "-af", "loudnorm=I=-14:LRA=11:TP=-1.5:print_format=json", "-f", "null", null_sink])
        .output()
        .map_err(|e| format!("Paso 1 Error: {}", e))?;

    let log = String::from_utf8_lossy(&pass1_output.stderr);
    let i = extract_json_value(&log, "input_i").ok_or("input_i missing".to_string())?;
    let lra = extract_json_value(&log, "input_lra").ok_or("input_lra missing".to_string())?;
    let tp = extract_json_value(&log, "input_tp").ok_or("input_tp missing".to_string())?;
    let thresh = extract_json_value(&log, "input_thresh").ok_or("input_thresh missing".to_string())?;

    let temp_path = input.with_file_name("temp_dsp_norm.mp3");
    let pass2_filter = format!("loudnorm=I=-14:LRA=11:TP=-1.5:measured_I={}:measured_LRA={}:measured_TP={}:measured_thresh={}:linear=true", i, lra, tp, thresh);

    let pass2_output = Command::new(get_ffmpeg_path())
        .args(["-y", "-i", input.to_str().unwrap(), "-af", &pass2_filter, "-c:a", "libmp3lame", "-q:a", "2", temp_path.to_str().unwrap()])
        .output()
        .map_err(|e| format!("Paso 2 Error: {}", e))?;

    if pass2_output.status.success() {
        atomic_replace(&temp_path, input)?;
        Ok(true)
    } else {
        if temp_path.exists() { let _ = fs::remove_file(temp_path); }
        Err(String::from_utf8_lossy(&pass2_output.stderr).to_string())
    }
}

pub async fn process_full_pipeline(input_path: String) -> Result<bool, String> {
    let input = Path::new(&input_path);
    if !input.exists() { return Err("Archivo no encontrado en I/O.".to_string()); }
    let temp_path = input.with_file_name("temp_dsp_full.mp3");

    let filter = "loudnorm=I=-14:LRA=11:TP=-1.5,silenceremove=start_periods=1:start_duration=0.05:start_threshold=-30dB,areverse,silenceremove=start_periods=1:start_duration=0.05:start_threshold=-30dB,areverse";

    let output = Command::new(get_ffmpeg_path())
        .args(["-y", "-i", input.to_str().unwrap(), "-af", filter, "-c:a", "libmp3lame", "-b:a", "320k", temp_path.to_str().unwrap()])
        .output()
        .map_err(|e| format!("OS Error: {}", e))?;

    if output.status.success() {
        atomic_replace(&temp_path, input)?;
        Ok(true)
    } else {
        if temp_path.exists() { let _ = fs::remove_file(temp_path); }
        Err(String::from_utf8_lossy(&output.stderr).to_string())
    }
}

pub async fn clear_watermark(input_path: String) -> Result<bool, String> {
    let input = Path::new(&input_path);
    let temp_path = input.with_file_name("temp_clear_meta.mp3");

    let output = Command::new(get_ffmpeg_path())
        .args(["-y", "-i", input.to_str().unwrap(), "-map", "0", "-c", "copy", "-metadata", "DjStudio_M3=", "-metadata", "DjStudio_M3_V2=", temp_path.to_str().unwrap()])
        .output()
        .map_err(|e| format!("OS Error: {}", e))?;

    if output.status.success() {
        atomic_replace(&temp_path, input)?;
        Ok(true)
    } else {
        if temp_path.exists() { let _ = fs::remove_file(temp_path); }
        Err(String::from_utf8_lossy(&output.stderr).to_string())
    }
}

pub async fn read_audio_genre(input_path: String) -> String {
    let output = Command::new(get_ffmpeg_path())
        .args(["-i", &input_path, "-f", "ffmetadata", "-"])
        .output();

    if let Ok(out) = output {
        let log = format!("{}{}", String::from_utf8_lossy(&out.stderr), String::from_utf8_lossy(&out.stdout));
        for line in log.to_lowercase().lines() {
            if line.starts_with("genre=") {
                return line.replace("genre=", "").trim().to_string();
            }
        }
    }
    "desconocido".to_string()
}